### ISSUE TITLE
This PR leaves the user in an odd state, giving both RS0016 and RS0037 errors when the `PublicApi.*.txt` files do not exist. Suggest the RS0016 fixer should include the required `#nullable enable` line to avoid fixing the missing files twice.

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn-analyzers#3957]
This PR leaves the user in an odd state, giving both RS0016 and RS0037 errors when the `PublicApi.*.txt` files do not exist. Suggest the RS0016 fixer should include the required `#nullable enable` line to avoid fixing the missing files twice.

_Originally posted by @dougbu in https://github.com/dotnet/roslyn-analyzers/pull/3916#issuecomment-669637958_

**Comment by dougbu** [comment]
Turns out the problem is worse: RS0037 issues **must** be fixed first or the RS0016 fixer will write out the `PublicApi.*.txt` files without the correct nullability annotations. I had to go back and remove the generated files then run the fixers in the supported order.

**Comment by dougbu** [comment]
Shoot, how would I run the RS0037 fixer before the RS0016 one❔

**Comment by mavasani** [comment]
@dougbu a better approach is just disable RS0037 for the repo through editorconfig entry “dotnet_diagnostic.RS0037.severity = none”, fix all RS0016 with a FixAll operation, then re-enable RS0037 by removing the above entry and doing a FixAll for it as a second step.

**Comment by mavasani** [comment]
https://docs.microsoft.com/en-us/visualstudio/code-quality/use-roslyn-analyzers?view=vs-2019#set-rule-severity-in-an-editorconfig-file for setting rule severity in editorconfig to disable a rule.

**Comment by mavasani** [comment]
Note RS0016 code fix cannot add nullable enable annotations by default- not everyone has nullable enabled entire projects or may not want to add nullability annotations to the public API files from scratch. RS0037 fixes should be done as a second step.

**Comment by dougbu** [comment]
1. The fixers should be able to detect what's enabled because nullability metadata appears in the assemblies in most cases.
2. It doesn't matter whether the RS0037 is enabled or not, the RS0016 fixer generates files that won't be valid for most of dotnet/aspnetcore.
    - The only fixer order that works is RS0016 ➡️ RS0037 ➡️ RS0036 and even that has bugs e.g. the RS0037 fixer adds duplicate `#nullable enable` lines to some projects when applied to an entire solution i.e. causes RS0024 errors.
    - There doesn't seem to be an RS0024 fixer.
    - Worse, at most three RS0024 errors are displayed after each build in VS.
3. Still getting the following error with no fixer. The `Enumerator` in question is a `struct`.
    ``` text
    Error	RS0016	Symbol 'implicit constructor for 'Enumerator'' is not part of the declared API.	Microsoft.AspNetCore.Routing.Abstractions C:\dd\dnx\aspnetcore\src\Http\Routing.Abstractions\src\Properties\AssemblyInfo.cs	12	Active
    ```
    - this might be an unrelated issue but the RS0016 fixer isn't offered and (of course) only the `#pragma` approach works in the AssemblyInfo.cs because there's nothing in the `Enumerator` source to annotate.

**Comment by mavasani** [comment]
> The fixers should be able to detect what's enabled because nullability metadata appears in the assemblies in most cases.

Analyzer/code fix do not have access to these project properties. This support was recently added to compiler in VS2019 16.8, but will take some time to be available for public API analyzer that is based of a much earlier version of compiler API. Can you file a separate issue for this feature request?

> It doesn't matter whether the RS0037 is enabled or not, the RS0016 fixer generates files that won't be valid for most of dotnet/aspnetcore.

 I was able to perform the following steps of fix all RS0016, RS0037 and RS0036 violations.

1. Install the public API analyzer on whole solution
2. Apply RS0016 fix all for solution. This should add public API files with non-nullable annotated entries
3. Close and re-open solution (this seems to be an unrelated Roslyn bug where adding new additional files to project is not re-creating analyzer options)
4. Verify RS0037 diagnostics to suggest adding nullable enable at top of public API files. Apply RS0037 fix all for solution. This should add `# nullable enabled` at top of all shipped api files in solution.
5. Verify RS0036 diagnostics in entire solution. Apply RS0036 fix all for solution.

> RS0037 fixer adds duplicate #nullable enable lines to some projects when applied to an entire solution i.e. causes RS0024 errors.

Duplicate entries is just a bug that I plan to fix. `RS0024` is generic diagnostic about invalid or unexpected entries in public API files when manually edited, and it is not feasible to fix majority of cases through an automated code fix.

> Still getting the following error with no fixer. The Enumerator in question is a struct

Yes, please file a separate issue.

**Comment by mavasani** [comment]
> RS0037 fixer adds duplicate #nullable enable lines to some projects when applied to an entire solution i.e. causes RS0024 errors.

@dougbu I wasn't to reproduce this. Can you please file a separate issue with repro steps to help reproduce this?

**Comment by mavasani** [comment]
@dougbu I can try to repro on your repo if you can let me know where to install the below SDK: https://github.com/dotnet/aspnetcore/blob/ed5034eb8ecde1e0151f70da43203affd1375c46/global.json#L3

**Comment by dougbu** [comment]
@mavasani run `.\restore.cmd` in the repo root to install the SDK, needed runtimes, and restore all projects before starting Visual Studio. Use the `.\startvs.cmd` with AspNetCore.sln or any `*.slnf` file to ensure VS uses the correct SDK.

**Comment by dougbu** [comment]
I've filed all the issues @mavasani #3963 and #3965 include more detailed repro steps than anything above.

**Comment by mavasani** [comment]
Closing this out in favor of individual issues.
